sw/device/tests/dif/dif_gpio_test.cc

```diff
@@ -0,0 +1,420 @@
+            kDifGpioResultOK);
+
+  EXPECT_WRITE32(GPIO_MASKED_OE_LOWER_REG_OFFSET, 0xCDCD3322);
+  EXPECT_EQ(dif_gpio_output_mode_masked_set(&gpio_, 0x0000CDCD, 0x55443322),
+            kDifGpioResultOK);
+
+  EXPECT_WRITE32(GPIO_MASKED_OE_UPPER_REG_OFFSET, 0xABAB5544);
+  EXPECT_EQ(dif_gpio_output_mode_masked_set(&gpio_, 0xABAB0000, 0x55443322),
+            kDifGpioResultOK);
+}
+
+// Input noise filter tests
+class InputFilterTest : public DifGpioTestWithDevice {};
+
+TEST_F(InputFilterTest, NullArgs) {
+  EXPECT_EQ(dif_gpio_input_noise_filter_masked_enable(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_input_noise_filter_masked_disable(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+}
+
+TEST_F(InputFilterTest, MaskedEnable) {
+  const uint32_t val = 0xABABABAB;
+  EXPECT_READ32(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, 0x0);
+  EXPECT_WRITE32(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, val);
+
+  EXPECT_EQ(dif_gpio_input_noise_filter_masked_enable(&gpio_, val),
+            kDifGpioResultOK);
+}
+
+TEST_F(InputFilterTest, MaskedDisable) {
+  const uint32_t val = 0xABABABAB;
+  EXPECT_READ32(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, 0xFFFFFFFF);
+  EXPECT_WRITE32(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, ~val);
+
+  EXPECT_EQ(dif_gpio_input_noise_filter_masked_disable(&gpio_, val),
+            kDifGpioResultOK);
+}
+
+// IRQ tests
+class IrqTest : public DifGpioTestWithDevice {};
+
+TEST_F(IrqTest, NullArgs) {
+  EXPECT_EQ(dif_gpio_irq_pin_test(nullptr, 0), kDifGpioResultInvalidArgument);
+
+  uint32_t out_arg_uint32_t;
+  EXPECT_EQ(dif_gpio_irq_all_read(nullptr, &out_arg_uint32_t),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_irq_all_read(&gpio_, nullptr),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_irq_all_read(nullptr, nullptr),
+            kDifGpioResultInvalidArgument);
+
+  bool out_arg_bool;
+  EXPECT_EQ(dif_gpio_irq_pin_read(nullptr, 0, &out_arg_bool),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_irq_pin_read(&gpio_, 0, nullptr),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_irq_pin_read(nullptr, 0, nullptr),
+            kDifGpioResultInvalidArgument);
+
+  EXPECT_EQ(dif_gpio_irq_pin_clear(nullptr, 0), kDifGpioResultInvalidArgument);
+
+  EXPECT_EQ(dif_gpio_irq_masked_enable(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+
+  EXPECT_EQ(dif_gpio_irq_masked_disable(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+
+  EXPECT_EQ(dif_gpio_irq_trigger_masked_disable(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+
+  EXPECT_EQ(dif_gpio_irq_trigger_masked_config(nullptr, 0xFFFFFFFF,
+                                               kDifGpioIrqEdgeRising),
+            kDifGpioResultInvalidArgument);
+}
+
+TEST_F(IrqTest, ForceSinglePin) {
+  for (uint32_t pin = 0; pin < 32; ++pin) {
+    EXPECT_WRITE32(GPIO_INTR_TEST_REG_OFFSET, {{pin, 1}});
```

Comment:
Nit: `1` -> `true` for single bit?

Reply:
I've been using `0x1` for single bits, actually. I'm not sure I'm fully comfortable with using `bool` here, because it's not a simple typedef in C++ as it is in C. I think this is more of a general test-style issue we should look into once more DIF tests exist, though.

Reply:
@silvestrst, while what you are suggesting is perfectly fine (and defined by the std), I prefer to keep `1` because it is what the function is literally expected to do.

@mcy, I initially preferred `1` to `0x1` because it is shorter, but using hex literals for `mock_mmio` can actually make the tests more readable. I'm on the fence. Let me know if you feel strongly about this and I'm happy to change all instances.